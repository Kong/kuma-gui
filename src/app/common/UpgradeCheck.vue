<template>
  <DataSource
    v-slot="{ data }"
    :src="`/control-plane/version/latest`"
  >
    <KAlert
      v-if="data && env('KUMA_VERSION') !== data.version"
      data-testid="upgrade-check"
      class="upgrade-check-alert"
      appearance="info"
      size="small"
    >
      <template #alertMessage>
        <div class="alert-content">
          <div>
            {{ t('common.product.name') }} update available
          </div>

          <div>
            <KButton
              appearance="primary"
              :to="t('common.product.href.install')"
            >
              Update
            </KButton>
          </div>
        </div>
      </template>
    </KAlert>
  </DataSource>
</template>

<script lang="ts" setup>
import { useEnv, useI18n } from '@/utilities'
const env = useEnv()
const { t } = useI18n()
</script>

<style lang="scss" scoped>
.k-alert.small {
  // Uses smaller paddings for this particular alert.
  padding: $kui-space-20 $kui-space-40;
}

.alert-content {
  display: flex;
  align-items: center;
  font-size: $kui-font-size-30;

  > *:first-of-type {
    margin-right: $kui-space-50;
  }
}
</style>
