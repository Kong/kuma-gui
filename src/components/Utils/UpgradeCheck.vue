<template>
  <div class="upgrade-check">
    <KAlert
      v-if="showNotice"
      appearance="warning"
      size="small"
    >
      <template slot="alertMessage">
        <div class="alert-content">
          <div>
            {{ tagline }} update available
          </div>
          <div>
            <KButton
              class="warning-button"
              appearance="primary"
              size="small"
              :to="url"
            >
              Update
            </KButton>
          </div>
        </div>
      </template>
    </KAlert>
  </div>
</template>

<script>
import axios from 'axios'
import compare from 'semver-compare'
import { mapGetters } from 'vuex'

export default {
  name: 'UpgradeCheck',
  data () {
    return {
      url: `${process.env.VUE_APP_INSTALL_URL}${process.env.VUE_APP_UTM}`,
      latestVerSrc: process.env.VUE_APP_VERSION_URL,
      latestVer: null,
      showNotice: false
    }
  },
  computed: {
    ...mapGetters({
      currentVer: 'config/getVersion',
      tagline: 'config/getTagline'
    })
  },
  beforeMount () {
    this.checkVersion()
  },
  methods: {
    checkVersion () {
      axios.get(this.latestVerSrc)
        .then(response => {
          const status = response.status
          const data = response.data

          if (status === 200 && data && data.length > 0) {
            this.latestVer = data
          }
        })
        .catch(error => {
          this.showNotice = false

          console.error(error)
        })
        .finally(() => {
          if (this.latestVer) {
            // compare the latest version to the currently running version
            // but only if we were able to set the latest version in the first place.
            const comparison = compare(this.latestVer, this.currentVer)

            if (comparison === 1) {
              this.showNotice = true
            } else {
              this.showNotice = false
            }
          } else {
            const timespan = 3 // months
            const today = new Date()
            const refDate = new Date('2020-06-03 12:00:00')
            const later = new Date(
              refDate.getFullYear(),
              refDate.getMonth() + timespan,
              refDate.getDate()
            )

            // compare dates and handle the notice accordingly
            if (today.getTime() >= later.getTime()) {
              this.showNotice = true
            } else {
              this.showNotice = false
            }
          }
        })
    }
  }
}
</script>

<style lang="scss" scoped>
.upgrade-check {

  .k-alert.k-alert {
    padding: var(--spacing-xxs) var(--spacing-xs);
  }
}

.alert-content {
  display: flex;
  align-items: center;
  font-size: var(--type-sm);

  > *:first-of-type {
    margin-right: var(--spacing-sm);
  }

  .k-button {
    text-decoration: none;

    &:after {
      display: none;
    }
  }
}

.warning-button {
--KButtonPrimaryBase: #f2a230;
--KButtonPrimaryHover: #f2a230;
--KButtonPrimaryActive: #f2a230;
}

</style>
