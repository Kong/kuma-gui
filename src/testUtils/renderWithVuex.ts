import { render, RenderOptions, RenderResult } from '@testing-library/vue'
import merge from 'lodash/merge'
import { StoreOptions } from 'vuex'
import Vue, { ComponentOptions } from 'vue'
import Store from '@/store'

function renderWithVuex(
  Component: typeof Vue | ComponentOptions<Vue>,
  options: RenderOptions<Vue> = { store: {}, routes: [] },
): RenderResult {
  const store: StoreOptions<any> = Store()

  // Render the component and merge the original store and the custom one
  // provided as a parameter. This way, we can alter some behaviors of the
  // initial implementation.

  return render(Component, { ...options, store: merge(store, options.store), routes: options.routes })
}

export default renderWithVuex
