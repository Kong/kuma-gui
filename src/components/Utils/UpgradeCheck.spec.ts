import renderWithVuex from '@/testUtils/renderWithVuex'
import UpgradeCheck from './UpgradeCheck.vue'

describe('UpgradeCheck.vue', () => {
  it('renders snapshot', async () => {
    const { container, findByText } = renderWithVuex(UpgradeCheck, {
      store: { modules: { config: { state: { kumaDocsVersion: '1.2.0', tagline: 'Kuma' } } } },
    })

    await findByText('Update')
    expect(container).toMatchSnapshot()
  })
})
