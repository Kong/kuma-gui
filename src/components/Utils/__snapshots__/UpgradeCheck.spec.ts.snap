// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpgradeCheck.vue renders snapshot 1`] = `
<div>
  <div
    class="upgrade-check"
  >
    <div
      class="k-alert warning small none"
      data-v-825647a6=""
      role="alert"
    >
      <!---->
      <!---->
      <div
        class="k-alert-msg-text"
        data-v-825647a6=""
      >
        <!---->
        <div
          class="k-alert-msg"
          data-v-825647a6=""
        >
          <div
            class="alert-content"
            data-v-825647a6=""
          >
            <div
              data-v-825647a6=""
            >
              
          Kuma update available
        
            </div>
             
            <div
              data-v-825647a6=""
            >
              <a
                class="k-button warning-button small rounded primary"
                data-v-6ae23500=""
                data-v-825647a6=""
                href="https://kuma.io/install/latest/?utm_source=Kuma&utm_medium=Kuma-GUI"
                type="button"
              >
                <!---->
                
            Update
          
                <!---->
              </a>
            </div>
          </div>
        </div>
        <!---->
      </div>
      <!---->
      <!---->
    </div>
  </div>
</div>
`;
